cranelift/codegen/src/machinst/buffer.rs

```diff
@@ -1774,16 +1774,79 @@ impl<T: CompilePhase> MachBufferFinalized<T> {
-            let range = usize::try_from(range.start).unwrap()..usize::try_from(range.end).unwrap();
+            let handler_range = call_site.exception_handler_range.clone();
+            let handler_range = usize::try_from(handler_range.start).unwrap()
+                ..usize::try_from(handler_range.end).unwrap();
             FinalizedMachCallSite {
                 ret_addr: call_site.ret_addr,
-                exception_handlers: &self.exception_handlers[range],
+                exception_handlers: &self.exception_handlers[handler_range],
             }
         })
     }
 }
 
+/// An item in the exception-handler list for a callsite, with label
+/// references.  Items are interpreted in left-to-right order and the
+/// first match wins.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum MachExceptionHandler {
+    /// A specific tag (in the current dynamic context) should be
+    /// handled by the code at the given offset.
+    Tag(ExceptionTag, MachLabel),
+    /// All exceptions should be handled by the code at the given
+    /// offset.
+    Default(MachLabel),
+    /// The dynamic context for interpreting tags is updated to the
+    /// value stored in the given machine location (in this frame's
+    /// context).
+    Context(ExceptionContextLoc),
+}
+
+impl MachExceptionHandler {
+    fn finalize<F: Fn(MachLabel) -> CodeOffset>(self, f: F) -> FinalizedMachExceptionHandler {
+        match self {
+            Self::Tag(tag, label) => FinalizedMachExceptionHandler::Tag(tag, f(label)),
+            Self::Default(label) => FinalizedMachExceptionHandler::Default(f(label)),
+            Self::Context(loc) => FinalizedMachExceptionHandler::Context(loc),
+        }
+    }
+}
+
+/// An item in the exception-handler list for a callsite, with final
+/// (lowered) code offsets. Items are interpreted in left-to-right
+/// order and the first match wins.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(
+    feature = "enable-serde",
+    derive(serde_derive::Serialize, serde_derive::Deserialize)
+)]
+pub enum FinalizedMachExceptionHandler {
+    /// A specific tag (in the current dynamic context) should be
+    /// handled by the code at the given offset.
+    Tag(ExceptionTag, CodeOffset),
+    /// All exceptions should be handled by the code at the given
+    /// offset.
+    Default(CodeOffset),
+    /// The dynamic context for interpreting tags is updated to the
+    /// value stored in the given machine location (in this frame's
+    /// context).
+    Context(ExceptionContextLoc),
+}
+
+/// A location for a dynamic exception context value.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(
+    feature = "enable-serde",
+    derive(serde_derive::Serialize, serde_derive::Deserialize)
+)]
+pub enum ExceptionContextLoc {
+    /// An offset from SP at the callsite.
+    SPOffset(u32),
+    /// A GPR at the callsite. The physical register number for the
+    /// GPR register file on the target architecture is used.
```

Comment:
Is this using the same numbering scheme as DWARF debuginfo? If so can this reuse the enum we use for variable location debuginfo? If not what numbering scheme is used?

Reply:
(err, [here](https://github.com/bytecodealliance/wasmtime/blob/88079b4f3f47e11c1263bde10b9c30bcbe86cc86/cranelift/codegen/src/isa/x64/inst/regs.rs#L15), rather)

Reply:
Would it be possible to use a literal `PReg` here? That way embedders could remap if they like, but there's a canonical definition for each register (the various backend helpers). Otherwise I'd agree it's not worth too much infrastructure just yet and it's ok to assert with a todo that a register doesn't pop out

Reply:
Perhaps; but we don't export regalloc2 types from the public API anywhere else, nor those backend helpers? The latter especially mean that the `PReg` isn't much more meaningful than the raw `hw_enc` -- that's the integer we are returning currently.

In fullness of time this seems like something an assembler layer should own -- canonical register names and ISA-defined numbering, then we can use this mapping in our public APIs -- but I guess no matter which type we wrap around that index today, the issue is that we don't have any public API surface that defines it.

Reply:
Ok makes sense, personally I'm ambivalent about leaving this as `u8` or just removing the variant for now, could go either way.